d2/src/main/proto/XdsD2.proto

```diff
@@ -30,15 +31,80 @@ message Stat {
 }
 
 message D2Node {
+  // Deprecated in favor of Node
+  option deprecated = true;
   Stat stat = 1;
   google.protobuf.Struct data = 2;
 }
 
 message D2SymlinkNode {
+  // Deprecated in favor of Node
+  option deprecated = true;
   Stat stat = 1;
   string masterClusterNodePath = 2;
 }
 
 message D2NodeMap {
+  // Deprecated in favor of D2UriMap
+  option deprecated = true;
   map<string, D2Node> nodes = 1;
 }
+
+message Node {
+  Stat stat = 1;
+  bytes data = 2 ;
+}
+
+// D2URI is a proto representation of com.linkedin.d2.balancer.properties.UriProperties. Note that a D2 UriProperties is
+// is designed to hold all the announcements of a cluster, which is why it's represented as a map of URI to data. The
+// UriProperties class is reused wholesale for serialization to write the data to ZK, which is why all fields are
+// actually maps, even though these maps only ever have one key in them. It is clear from the implementation of
+// ZooKeeperServer and ZooKeeperAnnouncer that there cannot ever be more than one URI in one ZK announcement, therefore
+// this new proto representation does not need to share the same shortcomings and can, instead, represent things more
+// linearly.
+//
+// Here is a sample ZK announcement in JSON serialized from a UriProperties for additional clarity on the fields that
+// are represented as maps when they do not need to be:
+//	{
+//		"weights": {
+//			"https://foo.stg.linkedin.com:18792/Toki/resources": 1.0
+//		},
+//		"partitionDesc": {
+//			"https://foo.stg.linkedin.com:18792/Toki/resources": {
+//				"0": {
+//					"weight": 1.0
+//				}
+//			}
+//		},
+//		"uriSpecificProperties": {
+//			"https://foo.stg.linkedin.com:18792/Toki/resources": {
+//				"com.linkedin.app.version": "0.1.76"
+//			}
+//		},
+//		"clusterName": "Toki"
+//	}
+message D2URI {
+  // The version of this announcement. When coming from ZK, this will be the node's mzxid.
+  int64 version = 1;
+
+  // The time at which this announcement was last updated. When coming from ZK this will be the node's mtime.
+  google.protobuf.Timestamp modified_time = 2;
+
+  // The name of the cluster this announcement belongs to. This is inferred from the original "clusterName" field.
+  string cluster_name = 3;
+
+  // The URI for this announcement, i.e. the host, port and context path that requests should be sent to.
+  string uri = 4;
```

Comment:
this assumes there is only one uri in a zk uri node data, but in rare cases there could be multiple. I have a concern with creating this D2URI structure instead of using the exact same ZK uri node data, please see my comment on the counterpart indis observer PR. thanks.

Reply:
For posterity, we double checked that there is no way that there can be more than one announcement. Additionally, the new data type should be much, much cheaper to deserialize and provides enough benefit to warrant the switch from raw ZK data to the pre-processed version